maitake/src/task/id.rs

```diff
@@ -0,0 +1,82 @@
+use core::fmt;
+
+/// A unique identifier for a running [task].
+///
+/// A `TaskId` is an opaque object that uniquely identifies each [task]
+/// created during the lifetime of a program. `TaskIds`s are guaranteed not to
```

Comment:
Might wait to slightly weaken this claim? Very long-running hardware could potentially wrap a u32.

Reply:
This is the guarantee `std` makes for `ThreadId`s, and it does this by using a mutexed `u64` on platforms without 64-bit atomics. We could either weaken the guarantee, or use a similar approach to ensure that task IDs are always `u64`s. We currently assert that the task ID counter doesn't wrap, so if we think there's the potential of wrapping on 32-bit platforms, we probably should also remove that assertion: https://github.com/hawkw/mycelium/blob/ec15a80e99601a52e826536f0b6773ccd32a89ac/maitake/src/task/id.rs#L32

This is the real reason I liked using addrs for task IDs, it has weaker uniqueness guarantees because an address might be reused, but it _does_ guarantee that task IDs will never collide with the IDs of *other running tasks*, which is a concern if counters can wrap . Tasks with low IDs are spawned early in the program's lifetime, and might be more likely to be long running tasks, so if (say) task `1` is a long-lived task in a program that runs for a long time and spawns a bunch of short lived tasks constantly, if a 32-bit counter wraps, we might see ID `1` assigned to a second task _while_ the original task `1` is still running. I would personally prefer IDs that are always guaranteed to be unique regardless of word size, even if they're _not_ guaranteed to not be reused... :woman_shrugging: